### PR TITLE
LPS-73633 Implement map-common, map-openstreetmap and map-google-maps modules with Metal

### PR DESCRIPTION
**Pull request by p2kmgcl**

### CODE REVIEW

File: modules/apps/foundation/map/map-common/.jshintrc [hunk 1]
```diff
@@ -0,0 +1,31 @@
+{
```
Comment (jbalsas) [inline modules/apps/foundation/map/map-common/.jshintrc]: Could you try to add something like https://github.com/metal/metal-clay-components/blob/master/.eslintrc instead?

Comment (jbalsas) [inline modules/apps/foundation/map/map-common/.jshintrc]: See the dependencies in https://github.com/metal/metal-clay-components/blob/master/package.json

File: modules/apps/foundation/map/map-common/build.gradle [hunk 2]
```diff
@@ -0,0 +1,3 @@
+transpileJS {
+	srcIncludes = "**/*.es.js"
```
Comment (jbalsas) [inline modules/apps/foundation/map/map-common/build.gradle]: This should not be needed if we're calling files `.es.js`

File: modules/apps/foundation/map/map-common/package.json [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+		"gulp-metal": "^2.0.2",
+		"karma": "^1.3.0",
+		"karma-html2js-preprocessor": "^1.1.0",
+		"liferay-karma-config": "0.0.4",
+		"liferay-module-config-generator": "^1.2.1",
+		"metal": "^2.5.12",
+		"metal-cli": "^4.0.1",
+		"metal-component": "^2.5.13",
+		"metal-dom": "^2.5.13",
+		"metal-events": "^2.5.16",
```
Comment (jbalsas) [inline modules/apps/foundation/map/map-common/package.json]: Some of these are probably just dependencies... `metal`, `metal-component`, `metal-dom`, `metal-events`?

File: modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
+import State from 'metal-state/src/State';
```
Comment (jbalsas) [inline modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js]: This should be fine with just `import State from `metal-state`

File: modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+import { isMap } from './validators.es';
+
+/**
+ * GeoJSONBase
+ * @abstract
```
Comment (jbalsas) [inline modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js]: Some more information about what this abstract base does?

File: modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+ */
+class GeoJSONBase extends State {
+	/**
+	 * Receives an object with native features data and tries
+	 * to parse it with the implemented method _getNativeFeatures.
+	 * If the generated Array of native features is not empty, it fires
+	 * a 'featuresAdded' event.
+	 */
```
Comment (jbalsas) [inline modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js]: Missing `@param`

File: modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+				}
+			);
+		}
+	}
+
+	/**
+	 * Callback executed when a native feature has been clicked.
+	 * It receives the feature as parameter, and emits a 'featureClick'
+	 * event with the wrapped feature as event data.
+	 * @private
```
Comment (jbalsas) [inline modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js]: Missing `@param`

File: modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+	/**
+	 * Parses a nativeFeaturesData object and return an array of the
+	 * parsed features. If no feature has been parsed it may return an
+	 * empty array.
+	 * @private
+	 * @abstract
+	 * @param {Object} nativeFeaturesData
+	 * @return {Object[]} List of native features to be added
+	 */
+	_getNativeFeatures(/* nativeFeaturesData */) {
```
Comment (jbalsas) [inline modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js]: I'm not sure about this docs... but I guess we could leave them for now... 🤔

File: modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+	 * Map to be used
+	 * @type {Object}
+	 */
+	map: {
+		validator: isMap
+	}
+};
+
+window.Liferay = window.Liferay || {};
+window.Liferay.MapGeojsonBase = GeoJSONBase;
```
Comment (jbalsas) [inline modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js]: I think we should move this somewhere else, even if that means a breaking change. We should avoid side-effects and global leaking

File: modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/MapBase.es.js [hunk 10]
```diff
@@ -0,0 +1,656 @@
+'use strict';
+
+import State from 'metal-state/src/State';
+import core from 'metal/src/core';
+import { buildFragment } from 'metal-dom/src/domNamed';
```
Comment (jbalsas) [inline modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/MapBase.es.js]: This should all be just `import State form 'metal-state'`, `import core from 'metal'`...

File: modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/MapBase.es.js [hunk 11]
```diff
@@ -0,0 +1,656 @@
...
+ * Object that will hold callbacks waiting for being executed
+ * when maps are created.
+ * @see MapBase.register()
+ * @see MapBase.get()
+ */
+const pendingCallbacks = {};
+
+/**
+ * MapBase
+ * @abstract
```
Comment (jbalsas) [inline modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/MapBase.es.js]: Some more info about what this abstract class does?

File: modules/apps/foundation/map/map-common/build.gradle (hunk 2 above)
Comment (p2kmgcl) [inline modules/apps/foundation/map/map-common/build.gradle]: :+1:

File: modules/apps/foundation/map/map-common/.jshintrc (hunk 1 above)
Comment (p2kmgcl) [inline modules/apps/foundation/map/map-common/.jshintrc]: :+1:

File: modules/apps/foundation/map/map-common/package.json (hunk 3 above)
Comment (p2kmgcl) [inline modules/apps/foundation/map/map-common/package.json]: :+1:

File: modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js (hunk 4 above)
Comment (p2kmgcl) [inline modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js]: :+1:

File: modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js (hunk 7 above)
Comment (p2kmgcl) [inline modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js]: :+1:

File: modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js (hunk 8 above)
Comment (p2kmgcl) [inline modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js]: :+1:

File: modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/MapBase.es.js (hunk 10 above)
Comment (p2kmgcl) [inline modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/MapBase.es.js]: :+1:

File: modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js (hunk 5 above)
Comment (p2kmgcl) [inline modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js]: :+1:

File: modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js (hunk 6 above)
Comment (p2kmgcl) [inline modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/GeoJSONBase.es.js]: :+1:

File: modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/MapBase.es.js (hunk 11 above)
Comment (p2kmgcl) [inline modules/apps/foundation/map/map-common/src/main/resources/META-INF/resources/js/MapBase.es.js]: :+1:

**Pull request closed without merging**